### DISCUSSION TITLE
上游出了个新的域名匹配机制

### DISCUSSION THREAD

**Discussion by CalmLong** [IrineSistiana/mosdns#93] [General]
有考虑加入吗

**Comment by IrineSistiana** [comment]
哪个上游? v2?

**Reply by CalmLong** [reply]
v2fly/v2ray-core/pull/587
v2fly/v2ray-core/pull/639

又改了

v2fly/v2ray-core/pull/743
v2fly/v2ray-core/pull/786

**Reply by IrineSistiana** [reply]
这算法666。原v2的匹配器导入二十万域名的话，能产生将近一百万个heap对象。流量大的时候可把gc累坏了。
这算法目测就产生几个。何止匹配速度提升25%。v2综合效率都能上几个百分点、

**Reply by CalmLong** [reply]
有考虑加入吗

**Reply by IrineSistiana** [reply]
有啊。但最近这几天没什么时间。未来几天也是。

**Reply by IrineSistiana** [reply]
额..... 这个新的 mph 匹配器....看不懂.
orz

**Reply by CalmLong** [reply]
其实现在的速度也挺快

**Reply by xczh** [reply]
抱歉啊刚看到这个discussion。原来上游已经有这个想法了，才看到 :) 通过前缀树来做fast match这是很自然的想法。考虑到全部重写还需要一点工作量，需要权衡投入-产出比。现在看来，可以把v2的`radix-tree`实现拿过来直接用 #100 
又要辛苦一下@IrineSistiana了。。虽然这个需求目前看起来优先级不算高

**Reply by IrineSistiana** [reply]
mosdns 的匹配器是仿着 v2ray 的写的。
https://github.com/v2fly/v2ray-core/blob/master/common/strmatcher/domain_matcher.go
(这个不是 radix-tree

如果用 radix-tree 的话，string 自带的 16 byte 开销太大了。说不定内存不降反而多了几倍。而且会产生非常多的堆对象。

个人觉得那个 AC 自动机性能最佳。就是看不懂.... v2fly/v2ray-core/pull/639

> 虽然这个需求目前看起来优先级不算高

确实匹配器确实已经足够快了。匹配一次 100ns。再优化也没什么空间了。(相比本地回环发一个 udp 包要 5000ns。开一个tls 连接要1e6ns。
